File: 2023/TypeScript/01.ts
import * as fs from "fs";
import * as path from "path";

function readInput(): string {
  const inputPath = path.join(__dirname, "../input/01.txt");
  return fs.readFileSync(inputPath, "utf8").trim();
}

function part1(input: string): number {
  return input.split("\r\n").reduce((sum, line) => {
    const digits = line.match(/\d/g) || ["0"];
    return sum + Number(digits[0] + digits.at(-1));
  }, 0);
}

function part2(input: string): number {
  const digitMap: { [key: string]: string } = {
    one: "1",
    two: "2",
    three: "3",
    four: "4",
    five: "5",
    six: "6",
    seven: "7",
    eight: "8",
    nine: "9",
  };

  return input.split("\r\n").reduce((sum, line) => {
    const matches = [
      ...line.matchAll(
        /(?=(\d|one|two|three|four|five|six|seven|eight|nine))/g
      ),
    ];
    const digits = matches.map((match) => digitMap[match[1]] || match[1]);

    return sum + Number(digits[0] + digits[digits.length - 1]);
  }, 0);
}

const input = readInput();

console.log("Part 1:", part1(input));
console.log("Part 2:", part2(input));

File: 2023/Rust/01.rs
mod day_01 {
    use std::fs;
    use std::path::{Path, PathBuf};

    fn read_input() -> String {
        let input_path: PathBuf = Path::new(file!()).parent().unwrap().join("../input/01.txt");
        fs::read_to_string(input_path)
            .expect("Failed to read input file")
            .trim()
            .to_string()
    }

    fn part1(input: &str) -> i32 {
        input.lines().map(|line| {
            let digits: Vec<char> = line.chars().filter(|c| c.is_ascii_digit()).collect();
            if digits.is_empty() {
                0
            } else {
                let first = digits[0];
                let last = digits[digits.len() - 1];
                format!("{}{}", first, last).parse::<i32>().unwrap_or(0)
            }
        }).sum()
    }

    fn part2(input: &str) -> i32 {
        input
        .lines()
        .filter_map(|line| {
            let line = line
                .replace("one", "o1e")
                .replace("two", "t2o")
                .replace("three", "t3e")
                .replace("four", "f4r")
                .replace("five", "f5e")
                .replace("six", "s6x")
                .replace("seven", "s7n")
                .replace("eight", "e8t")
                .replace("nine", "n9e");
            
            let mut digits = line.chars().filter(|c| c.is_ascii_digit());
            let first = digits.next()?;
            let last = digits.last().unwrap_or(first);
            
            Some(format!("{}{}", first, last).parse::<i32>().ok()?)
        })
        .sum()
    }

    pub fn run() {
        let input: String = read_input();

        println!("Part 1: {}", part1(&input));
        println!("Part 2: {}", part2(&input));
    }
}

pub use day_01::run;
